qubesansible/manage.py

```diff
@@ -0,0 +1,404 @@
+            dispvm.start()
+        return dispvm
+
+    @staticmethod
+    def ansible_to_native(var):
+        if isinstance(var, AnsibleBaseYAMLObject):
+            return var.__class__.__bases__[1](var)
+        return var
+
+    def run_play(self, play):
+        """Runs the given play on the mgmt dispvm of the host"""
+        self.vm = self.app.domains.get_blind(self.host_name)
+        if not self.vm:
+            display.display(f"Host {self.host} not found", "red")
+            return 0
+
+        dispvm = self._start_mgmt_disp_vm()
+
+        self._add_rpc_policies()
+        self.temp_dir.mkdir()
+
+        try:
+            self._add_play(play)
+            self._add_roles(play)
+            self._add_host_vars()
+            self._add_group_vars()
+            tar_file_path = self._build_tar()
+            ansible_args = self._build_ansible_args()
+
+
+            dispvm.run_service(
+                "qubes.Filecopy",
+                localcmd="/usr/lib/qubes/qfile-dom0-agent {}".format(
+                    tar_file_path
+                ),
+            ).wait()
+
+            p = dispvm.run_service("qubes.AnsibleVM")
+            (untrusted_stdout, untrusted_stderr) = p.communicate(
+                (
+                    tar_file_path.name
+                    + "\n"
+                    + self.host_name
+                    + "\n"
+                    + "\n".join(ansible_args)
+                    + "\n"
+                ).encode()
+            )
+            return (untrusted_stdout.decode("utf-8"),
+                    untrusted_stderr.decode("utf-8"),
+                    self.dispvm_mgmt_name,
+                    play.name)
+
+        finally:
+            self._remove_rpc_policies()
+            shutil.rmtree(self.temp_dir)
+            if not self.dispvm_initially_running:
+                dispvm.shutdown()
+                while dispvm.is_running():
+                    time.sleep(1)
+                time.sleep(2)
+
+
+def run_one(app, host, inventory, variable_manager, loader, play):
+    return ManageVM(
+        app, host, inventory, variable_manager, loader
+    ).run_play(play)
+
+
+class ManageVMRunner:
+    def __init__(
+        self,
+        app,
+        hosts,
+        inventory,
+        variable_manager,
+        loader,
+        max_concurrency=3,
+    ):
+        self.app = app
+        self.hosts = hosts
+        self.inventory = inventory
+        self.variable_manager = variable_manager
+        self.loader = loader
+        self.max_concurrency = max_concurrency
+
+    @staticmethod
+    def collect_result(result_tuple):
+        stdout, stderr, dispvm, play_name = result_tuple
+        display.banner(
+            f"QUBESOS [{dispvm}: PLAY {play_name}]"
+        )
+        if stderr:
+            display.display(stderr, 'red')
+        if stdout:
+            display.display(stdout, "bright blue")
+
+    @staticmethod
+    def collect_error(error):
+        display.display(error, 'red')
```

Comment:
This gets an exception object, not string. And `display.display()` is not happy about non-str.

Reply:
Oh, and when this line fails, the whole tool spins with 100% CPU and stays this way.